ptp4l.8
.TH PTP4l 8 "March 2024" "linuxptp"
.SH NAME
ptp4l - PTP Boundary/Ordinary/Transparent Clock

.SH SYNOPSIS
.B ptp4l
[
.B \-AEP246HSLmqsv
] [
.BI \-f " config"
] [
.BI \-p " phc-device"
] [
.BI \-l " print-level"
]
[
.BI \-i " interface"
] [
.I long-options
]
.I .\|.\|.

.SH DESCRIPTION
.B ptp4l
is an implementation of the Precision Time Protocol (PTP) according to IEEE
standard 1588 for Linux. It implements Boundary Clock (BC), Ordinary Clock
(OC), and Transparent Clock (TC).

.SH OPTIONS
.TP
.B \-A
Select the delay mechanism automatically. Start with E2E and switch to P2P when
a peer delay request is received.
.TP
.B \-E
Select the delay request-response (E2E) mechanism. This is the default
mechanism. All clocks on single PTP communication path must use the same
mechanism. A warning will be printed when a peer delay request is received on
port using the E2E mechanism.
.TP
.B \-P
Select the peer delay (P2P) mechanism. A warning will be printed when a delay
request is received on port using the P2P mechanism.
.TP
.B \-2
Select the IEEE 802.3 network transport.
.TP
.B \-4
Select the UDP IPv4 network transport. This is the default transport.
.TP
.B \-6
Select the UDP IPv6 network transport.
.TP
.B \-H
Select the hardware time stamping. All ports specified by the
.B \-i
option and in the configuration file must be attached to the same PTP hardware
clock (PHC). This is the default time stamping.
.TP
.B \-S
Select the software time stamping.
.TP
.B \-L
Select the legacy hardware time stamping.
.TP
.BI \-f " config"
Read configuration from the specified file. No configuration file is read by
default.
.TP
.BI \-i " interface"
Specify a PTP port, it may be used multiple times. At least one port must be
specified by this option or in the configuration file.
.TP
.BI \-p " phc-device"
(This option is deprecated.)
Before Linux kernel v3.5 there was no way to discover the PHC device
associated with a network interface.  This option specifies the PHC
device (e.g. /dev/ptp0) to be used when running on legacy kernels.
.TP
.B \-s
Enable the clientOnly mode.
.TP
.BI \-l " print-level"
Set the maximum syslog level of messages which should be printed or sent to
the system logger. The default is 6 (LOG_INFO).
.TP
.B \-m
Print messages to the standard output.
.TP
.B \-q
Don't send messages to the system logger.
.TP
.B \-v
Prints the software version and exits.
.TP
.BI \-h
Display a help message.

.SH LONG OPTIONS

Each and every configuration file option (see below) may also appear
as a "long" style command line argument.  For example, the clientOnly
option may be set using either of these two forms.

.RS
\f(CW\-\-clientOnly 1   \-\-clientOnly=1\fP
.RE

Option values given on the command line override values in the global
section of the configuration file.

.SH CONFIGURATION FILE

The configuration file is divided into sections. Each section starts with a
line containing its name enclosed in brackets and it follows with settings.
Each setting is placed on a separate line, it contains the name of the
option and the value separated by whitespace characters. Empty lines and lines
starting with # are ignored.

There are three different section types.

.TP
.B 1.
The global section (indicated as
.BR [global] )
sets the program options, clock options and default port options. Other
sections are port specific sections and they override the default port options.
.TP
.B 2.
Port sections give the name of the configured port (e.g.
.BR [eth0] ).
Ports specified in the configuration file don't need to be
specified by the
.B \-i
option. An empty port section can be used to replace the command line option.
.TP
.B 3.
Tables for configuring unicast discovery begin with
.B \%[unicast_master_table].

See UNICAST DISCOVERY OPTIONS, below.

.SH PORT OPTIONS

.TP
.B active_key_id
Used in conjunction with \fBspp\fR and \fBsa_file\fR directives to
specify which key from the \fBspp\fR defined Security Association
should be used for outbound icv calculations. All Security Associations
are read from the file specified by \fBsa_file\fR. Requires \fBspp\fR
and \fBsa_file\fR directives. Must be in the range of 1 to 2^32-1,
inclusive. The default is 0 (disabled).

.TP
.B allowedLostResponses
The number of missed peer delay responses before the asCapable variable is
reset.
The default is 3.

.TP
.B announceReceiptTimeout
The number of missed Announce messages before the last Announce messages
expires.
The default is 3.

.TP
.B boundary_clock_jbod
When running as a boundary clock (that is, when more than one network
interface is configured), ptp4l performs a sanity check to make sure
that all of the ports share the same hardware clock device. This
option allows ptp4l to work as a boundary clock using "just a bunch of
devices" that are not synchronized to each other. For this mode, the
collection of clocks must be synchronized by an external program, for
example phc2sys(8) in "automatic" mode.
The default is 0 (disabled).

.TP
.B cmlds.client_address
Specifies the source address for the UNIX domain socket that receives
peer delay measurement when using the "COMMON_P2P" delay mechanism.
The default is /var/run/cmlds_client.

.TP
.B cmlds.domainNumber
Specifies the domainNumber message field for communications with the
local Common Mean Link Delay Service, used with the "COMMON_P2P" delay
mechanism.
The default is 0.

.TP
.B cmlds.majorSdoId
Specifies the transportSpecific message field for communications with
the local Common Mean Link Delay Service, used with the "COMMON_P2P"
delay mechanism.
The default is 2.

.TP
.B cmlds.port
Specifies the port number of local Common Mean Link Delay Service from
which to accept delay measurements.  The range of valid port numbers
is one to 65535, but the special value "0" configures the port number
of the port requesting the measurements.
The default is 0, meaning the requesting port number.

.TP
.B
cmlds.server_address
Specifies the UNIX domain socket address of the local Common Mean Link
Delay Service, for use with the "COMMON_P2P" delay mechanism.
The default is /var/run/cmlds_server.

.TP
.B delayAsymmetry
The time difference in nanoseconds of the transmit and receive
paths. This value should be positive when the server-to-client
propagation time is longer and negative when the client-to-server time
is longer. The default is 0 nanoseconds.

.TP
.B delay_filter
Select the algorithm used to filter the measured delay and peer delay. Possible
values are moving_average and moving_median.
The default is moving_median.

.TP
.B delay_filter_length
The length of the delay filter in samples.
The default is 10.

.TP
.B delay_mechanism

Select the delay mechanism. Possible values are Auto, COMMON_P2P, E2E,
P2P, and NONE.
The default is E2E.

.TP
.B delay_request_variability
The variability of the interval between transmitted delay requests around the
minimum mean time interval specified by the \fBlogMinDelayReqInterval\fR
option. Must be in the range of 0.0 to 1.0.
The default is 1.0 (i.e. +/- 100% of the mean time interval).

.TP
.B delay_response_timeout
The number of delay response messages that may go missing before
triggering a synchronization fault. Setting this option to "0" will
disable the delay response timeout.
The default is 0 or disabled.

.TP
.B egressLatency
Specifies the difference in nanoseconds between the actual transmission
time at the reference plane and the reported transmit time stamp. This
value will be added to egress time stamps obtained from the hardware.
The default is 0.

.TP
.B fault_badpeernet_interval
The time in seconds between the detection of a peer network misconfiguration
and the fault being reset. The port is disabled for the duration of the
interval. The value is in seconds and the special key word ASAP will let
the fault be reset immediately.
The default is 16 seconds.

.TP
.B fault_reset_interval
The time in seconds between the detection of a port's fault and the fault
being reset. This value is expressed as a power of two. Setting this
value to \-128 or to the special key word "ASAP" will let the fault be
reset immediately.
The default is 4 (16 seconds).

.TP
.B follow_up_info
Include the 802.1AS data in the Follow_Up messages if enabled.
The default is 0 (disabled).

.TP
.B G.8275.portDS.localPriority
The Telecom Profiles (ITU-T G.8275.1 and G.8275.2) specify an
alternate Best Master Clock Algorithm (BMCA) with a unique data set
comparison algorithm.  The value of this option is associated with
Announce messages arriving on a particular port and is used as a tie
breaker whenever clockClass, clockAccuracy, offsetScaledLogVariance,
and priority2 are equal. This option is only used when
"dataset_comparison" is set to "G.8275.x".
The default value is 128.

Warning: the BMCA is guaranteed to produce a spanning tree (that is, a
timing network without loops) only when using the default values of
G.8275.defaultDS.localPriority and G.8275.portDS.localPriority.
Careful network engineering is needed when using non-default values.

.TP
.B hybrid_e2e
Enables the "hybrid" delay mechanism from the draft Enterprise
Profile. When enabled, ports in the client state send their delay
request messages to the unicast address taken from the server's
announce message. Ports in the server state will reply to unicast
delay requests using unicast delay responses. This option has no
effect if the delay_mechanism is set to P2P.
The default is 0 (disabled).

.TP
.B ignore_transport_specific
By default, incoming messages are dropped if their transportSpecific
field does not match the configured value.  However, many of
transports specified in the 1588 standard mandate ignoring this field.
Moreover, some equipment is known to set the reserved bits.
Configuring this option as 1 causes this field to be ignored
completely on receive.  The default is 0.

.TP
.B ingressLatency
Specifies the difference in nanoseconds between the reported receive
time stamp and the actual reception time at reference plane. This value
will be subtracted from ingress time stamps obtained from the hardware.
The default is 0.

.TP
.B inhibit_delay_req
Don't send any delay requests. This will need the asCapable config option to be
set to 'true'. This is useful when running as a designated server who does not
need to calculate offset from client. The default is 0 (disabled).

.TP
.B inhibit_multicast_service
Some unicast mode profiles insist that no multicast message are ever
transmitted.  Setting this option inhibits multicast transmission.
The default is 0 (multicast enabled).

.TP
.B logAnnounceInterval
The mean time interval between Announce messages. A shorter interval makes
ptp4l react faster to the changes in the client/server hierarchy. The interval
should be the same in the whole domain. It's specified as a power of two in
seconds.
The default is 1 (2 seconds).

.TP
.B logMinDelayReqInterval
The minimum permitted mean time interval between Delay_Req messages. A shorter
interval makes ptp4l react faster to the changes in the path delay. It's
specified as a power of two in seconds.
The default is 0 (1 second).

.TP
.B logMinPdelayReqInterval
The minimum permitted mean time interval between Pdelay_Req messages. It's
specified as a power of two in seconds.
The default is 0 (1 second).

.TP
.B logSyncInterval
The mean time interval between Sync messages. A shorter interval may improve
accuracy of the local clock. It's specified as a power of two in seconds.
The default is 0 (1 second).

.TP
.B masterOnly
This option is deprecated and will be removed in a future release.
Use "serverOnly" instead.

.TP
.B min_neighbor_prop_delay
Lower limit for peer delay in nanoseconds. If the estimated peer delay is
smaller than this value the port is marked as not 802.1AS capable.

.TP
.B neighborPropDelayThresh
Upper limit for peer delay in nanoseconds. If the estimated peer delay is
greater than this value the port is marked as not 802.1AS capable.

.TP
.B network_transport
Select the network transport. Possible values are UDPv4, UDPv6 and L2.
The default is UDPv4.

.TP
.B net_sync_monitor
Enables the NetSync Monitor (NSM) protocol. The NSM protocol allows a
station to measure how well another node is synchronized. The monitor
sends a unicast delay request to the node, which replies
unconditionally with unicast delay response, sync, and follow up
messages. If the monitor is synchronized to the GM, it can use the
time stamps in the message to estimate the node's offset.
The default is 0 (disabled).

.TP
.B operLogPdelayReqInterval
The Pdelay Request messages interval to be used once the clock enters
the SERVO_LOCKED_STABLE state.  If the 'msg_interval_request' option
is set, then the local client port will adopt this rate when the local
clock enters the "locked stable" state.  This option is specified as a
power of two in seconds, and the default value is 0 (1 second).

.TP
.B operLogSyncInterval
The Sync message interval to be requested once the clock enters the
SERVO_LOCKED_STABLE state.  If the 'msg_interval_request' option is
set, then the local client port will request the remote server to
switch to the given message rate via a signaling message containing a
Message interval request TLV.  This option is specified as a power of
two in seconds, and default value is 0 (1 second).

.TP
.B path_trace_enabled
Enable the mechanism used to trace the route of the Announce messages.
The default is 0 (disabled).

.TP
.B phc_index
Specifies the index of the PHC to be used for synchronization with hardware
timestamping. This option is useful with virtual clocks running on top of a
free-running physical clock (created by writing to
/sys/class/ptp/ptp*/n_vclocks).
The default is -1, which means the index will be set to the PHC associated with
the interface, or the device specified by the \fB-p\fP option.

.TP
.B power_profile.2011.grandmasterTimeInaccuracy
Specifies the time inaccuracy of the GM in nanoseconds.  Relevant only
when power_profile.version is 2011.  This value may be changed
dynamically using the POWER_PROFILE_SETTINGS_NP management message.
The default is -1 meaning unknown inaccuracy.

.TP
.B power_profile.2011.networkTimeInaccuracy
Specifies the time inaccuracy of the network in nanoseconds.  Relevant
only when power_profile.version is 2011.  This value may be changed
dynamically using the POWER_PROFILE_SETTINGS_NP management message.
The default is -1 meaning unknown inaccuracy.

.TP
.B power_profile.2017.totalTimeInaccuracy
Specifies the sum of the GM, network, and local node inaccuracies in
nanoseconds.  Relevant only when power_profile.version is 2017.  This
value may be changed dynamically using the POWER_PROFILE_SETTINGS_NP
management message.  The default is -1 meaning unknown inaccuracy.

.TP
.B power_profile.grandmasterID
Specifies an optional, non-zero identification code for the GM.  Note
that the code is an arbitrary, power profile specific integer, not
necessarily related to the clockIdentity in any way.  This value may
be changed dynamically using the POWER_PROFILE_SETTINGS_NP management
message.  The default is 0 meaning unused.

.TP
.B power_profile.version
Specifies the power profile version to be used.  Valid values are
"none", "2011", or "2017".
This value may be changed dynamically using the
POWER_PROFILE_SETTINGS_NP management message.
The default is "none".

.TP
.B profileIdentity
The profile identity is a 6-octet array in the clock description
referencing the PTP profile currently configured on the port. This value
should reflect the identifiers assigned to each profile. Allowed values
are of the form 00:00:00:00:00:00.
If the default 00:00:00:00:00:00 is used or if not set at all, the
profileIdentity will be detected from the configuration or set to the Default
profile (00:1B:19:00:01:00).
The default OUI value of 00:1B:19 represent the IEEE P1588 standard.

.TP
.B ptp_dst_ipv4
The IPv4 address to which PTP messages should be sent.
Relevant only with UDPv4 transport. The default is 224.0.1.129.
The default follows IEEE P1588, IPv4 multicast addresses, 'PTP-primary'.

.TP
.B p2p_dst_ipv4
The IPv4 address to which peer delay messages should be sent.
Relevant only with UDPv4 transport. The default is 224.0.0.107.
The default follows IEEE P1588, IPv4 multicast addresses, 'PTP-pdelay'.

.TP
.B ptp_dst_ipv6
The IPv6 address to which PTP messages should be sent.
The second byte of the address is substituted with udp6_scope.
Relevant only with UDPv6 transport. The default is FF0E:0:0:0:0:0:0:181.
The default follows IEEE P1588, IPv6 multicast addresses, 'PTP-primary' with 'Global scope'.

.TP
.B p2p_dst_ipv6
The IPv6 address to which peer delay messages should be sent.
Relevant only with UDPv6 transport. The default is FF02:0:0:0:0:0:0:6B.
The default follows IEEE P1588, IPv6 multicast addresses, 'PTP-pdelay'.

.TP
.B ptp_dst_mac
The MAC address to which PTP messages should be sent.
Relevant only with L2 transport. The default is 01:1B:19:00:00:00.
The default follows IEEE P1588, Multicast MAC addresses, 'All except peer-to-peer delay'.

.TP
.B p2p_dst_mac
The MAC address to which peer delay messages should be sent.
Relevant only with L2 transport. The default is 01:80:C2:00:00:0E.
The default follows IEEE P1588, Multicast MAC addresses, 'Peer-to-peer delay'.

.TP
.B serverOnly
Setting this option to one (1) prevents the port from entering the
client state. In addition, the local clock will ignore Announce
messages received on this port. This option's intended use is to
support the Telecom Profiles according to ITU-T G.8265.1, G.8275.1,
and G.8275.2. The default value is zero or false.

.TP
.B spp
Specifies the Security Parameters Pointer of the desired Security
Association to be used for Authentication TLV support for a given port.
Any port with an assigned spp will attach Authentication TLVs to all
outbound messages and check for Authentication TLVs on all inbound
messages in accordance to the corresponding security association
sourced via the \fBsa_file\fR directive. Outbound Authentication TLVs
are generated using the key specified by \fBactive_key_id\fR. Not
compatible with one step ports or advertised versions less then
PTPv2.1. Requires \fBsa_file\fR and \fBactive_key_id\fR directives.
Must be in the range of 0 to 255, inclusive.
The default is -1 (disabled).

.TP
.B syncReceiptTimeout
The number of sync/follow up messages that may go missing before
triggering a Best Master Clock election. This option is used for
running in gPTP mode according to the 802.1AS-2011 standard. Setting
this option to "0" will disable the sync message timeout.
The default is 0 or disabled.

.TP
.B transportSpecific
The transport specific field. Must be in the range 0 to 255.
The default is 0.

.TP
.B tsproc_mode
Select the time stamp processing mode used to calculate offset and delay.
Possible values are filter, raw, filter_weight, raw_weight. Raw modes perform
well when the rate of sync messages (logSyncInterval) is similar to the rate of
delay messages (logMinDelayReqInterval or logMinPdelayReqInterval). Weighting
is useful with larger network jitters (e.g. software time stamping).
The default is filter.

.TP
.B udp_ttl
Specifies the Time to live (TTL) value for IPv4 multicast messages and the hop
limit for IPv6 multicast messages. This option is only relevant with the IPv4
and IPv6 UDP transports. The default is 1 to restrict the messages sent by
.B ptp4l
to the same subnet.

.TP
.B unicast_listen
When enabled, this option allows the port to grant unicast message
contracts.  Incoming requests for will be granted limited only by the
amount of memory available.
The default is 0 (disabled).

.TP
.B unicast_master_table
When set to a positive integer, this option specifies the table id to
be used for unicast discovery.  Each table lives in its own section
and has a unique, positive numerical ID.  Entries in the table are a
pair of transport type and protocol address.
The default is 0 (unicast discovery disabled).

.TP
.B unicast_req_duration
The service time in seconds to be requested during unicast discovery.
Note that the remote node is free to grant a different duration.
The default is 3600 seconds or one hour.

.SH PROGRAM AND CLOCK OPTIONS

.TP
.B asCapable
If set to 'true', all the checks which can unset asCapable variable (as
described in Section 10.2.4.1 of 802.1AS) are skipped. If set to 'auto',
asCapable is initialized to 'false' and will be set to 'true' after the
relevant checks have passed. The default value is 'auto'.

.TP
.B assume_two_step
Treat one-step responses as two-step if enabled. It is used to work around
buggy 802.1AS switches.
The default is 0 (disabled).

.TP
.B BMCA
This option enables use of static roles for server and client devices
instead of running the best master clock algorithm (BMCA) described in
1588 profile. This can be used to speed up the start time for servers
and clients when you know the roles of the devices in advance.  When set to
\'noop', the traditional BMCA algorithm used by 1588 is skipped. masterOnly and
clientOnly will be used to determine the server or client role for the device. In a
bridge, clientOnly (which is a global option) can be set to make all ports
assume the client role. masterOnly (which is a per-port config option) can then
be used to set individual ports to take on the server role.
The default value is 'ptp' which runs the BMCA related state machines.

.TP
.B check_fup_sync
Because of packet reordering that can occur in the network, in the
hardware, or in the networking stack, a follow up message can appear
to arrive in the application before the matching sync message. As this
is a normal occurrence, and the sequenceID message field ensures
proper matching, the ptp4l program accepts out of order packets. This
option adds an additional check using the software time stamps from
the networking stack to verify that the sync message did arrive
first. This option is only useful if you do not trust the sequence IDs
generated by the server.
The default is 0 (disabled).

.TP
.B clientOnly
The local clock is a client-only clock if enabled. The default is 0 (disabled).

.TP
.B clockAccuracy
The clockAccuracy attribute of the local clock. It is used in the PTP server
selection algorithm.
The default is 0xFE.

.TP
.B clockClass
The clockClass attribute of the local clock. It denotes the traceability of the
time distributed by the grandmaster clock.
The default is 248.

.TP
.B clock_class_threshold
The maximum clock class value from master, acceptable to subordinate
clock beyond which it moves out of lock state.
The default value is 248.

.TP
.B clockIdentity
The clockIdentity attribute of the local clock.
The clockIdentity is an 8-octet array and should in this configuration be
written in textual form, see default. It should be unique since it is used to
identify the specific clock.
If default is used or if not set at all, the clockIdentity will be automatically
generated.
The default is "000000.0000.000000"

.TP
.B clock_servo
The servo which is used to synchronize the local clock. Valid values
are "pi" for a PI controller, "linreg" for an adaptive controller
using linear regression, "ntpshm" and "refclock_sock" for the NTP SHM and
chrony SOCK reference clocks respectively to allow another process to
synchronize the local clock, and "nullf" for a servo that always dials
frequency offset zero (for use in SyncE nodes).
The default is "pi."

.TP
.B clock_type
Specifies the kind of PTP clock.  Valid values are "OC" for ordinary
clock, "BC" for boundary clock, "P2P_TC" for peer to peer transparent
clock, and "E2E_TC" for end to end transparent clock.  An multi-port
ordinary clock will automatically be configured as a boundary clock.
The default is "OC".

.TP
.B dataset_comparison
Specifies the method to be used when comparing data sets during the
Best Master Clock Algorithm.  The possible values are "ieee1588" and
"G.8275.x".  The default is "ieee1588".

.TP
.B domainNumber
The domain attribute of the local clock.
The default is 0.

.TP
.B dscp_event
Defines the Differentiated Services Codepoint (DSCP) to be used for PTP
event messages. Must be a value between 0 and 63. There are several media
streaming standards out there that require specific values for this option.
For example 46 (EF PHB) in AES67 or 48 (CS6 PHB) in RAVENNA. The default
is 0.

.TP
.B dscp_general
Defines the Differentiated Services Codepoint (DSCP) to be used for PTP
general messages. Must be a value between 0 and 63. There are several media
streaming standards out there that recommend specific values for this option.
For example 34 (AF41 PHB) in AES67 or 46 (EF PHB) in RAVENNA. The default
is 0.

.TP
.B first_step_threshold
The maximum offset the servo will correct by changing the clock frequency (phase
when using nullf servo) instead of stepping the clock. This is only applied on
the first update. It's specified in seconds. When set to 0.0, the servo won't
step the clock on start.
The default is 0.00002 (20 microseconds).
This option used to be called
.BR pi_f_offset_const .

.TP
.B free_running
Don't adjust the local clock if enabled. The measured offset and frequency will
be printed at an interval limited by the \fBfreq_est_interval\fR option.
The default is 0 (disabled).

.TP
.B freq_est_interval
The time interval over which is estimated the ratio of the local and
peer clock frequencies. It is specified as a power of two in seconds.
The default is 1 (2 seconds).

.TP
.B G.8275.defaultDS.localPriority
The Telecom Profiles (ITU-T G.8275.1 and G.8275.2) specify an
alternate Best Master Clock Algorithm (BMCA) with a unique data set
comparison algorithm.  The value of this option is associated with the
local clock and is used as a tie breaker whenever clockClass,
clockAccuracy, offsetScaledLogVariance, and priority2 are equal. This
option is only used when "dataset_comparison" is set to "G.8275.x".
The default value is 128.

Warning: the BMCA is guaranteed to produce a spanning tree (that is, a
timing network without loops) only when using the default values of
G.8275.defaultDS.localPriority and G.8275.portDS.localPriority.
Careful network engineering is needed when using non-default values.

.TP
.B gmCapable
If this option is enabled, then the local clock is able to become grand master.
This is only for use with 802.1AS clocks and has no effect on 1588 clocks.
The default is 1 (enabled).

.TP
.B ignore_source_id
This will disable source port identity checking for Sync and Follow_Up
messages. This is useful when the announce messages are disabled in the server
and the client does not have any way to know the server's identity.
The default is 0 (disabled).

.TP
.B inhibit_announce
This will disable the timer for announce messages (i.e. FD_MANNO_TIMER) and
also the announce message timeout timer (i.e. FD_ANNOUNCE_TIMER). This is used
by the Automotive profile as part of switching over to a static BMCA.  If this
option is enabled, ignore_source_id has to be enabled in the client because it
has no way to identify the server in the Sync and Follow_Up messages. The
default is 0 (disabled).

.TP
.B initial_delay
The initial path delay of the clock in nanoseconds used for synchronization of
the clock before the delay is measured using the E2E or P2P delay mechanism. If
set to 0, the clock will not be updated until the delay is measured.
The default is 0.

.TP
.B interface_rate_tlv
When the client and server are operating at different interface rates, delay
asymmetry caused by different interface rates needs to be compensated for.
The server sends its interface rate using interface rate TLV
as per G.8275.2 Annex D.
The default is 0 (does not support interface rate TLV).

.TP
.B hwts_filter
Select the hardware time stamp filter setting mode.
Possible values are normal, check, full.
Normal mode set the filters as needed.
Check mode only check but do not set.
Full mode set the receive filter to mark all packets with hardware time stamp,
 so all applications can get them.
The default is normal.

.TP
.B kernel_leap
When a leap second is announced, let the kernel apply it by stepping the clock
instead of correcting the one-second offset with servo, which would correct the
one-second offset slowly by changing the clock frequency (unless the
.B step_threshold
option is set to correct such offset by stepping).
Relevant only with software time stamping. The default is 1 (enabled).

.TP
.B logging_level
The maximum logging level of messages which should be printed.
The default is 6 (LOG_INFO).

.TP
.B manufacturerIdentity
The manufacturer id which should be an OUI owned by the manufacturer.
The default is 00:00:00.

.TP
.B max_frequency
The maximum allowed frequency adjustment of the clock in parts per billion
(ppb). This is an additional limit to the maximum allowed by the hardware. When
set to 0, the hardware limit will be used.
The default is 900000000 (90%).
This option used to be called
.BR pi_max_frequency .

.TP
.B maxStepsRemoved
When using this option, if the value of stepsRemoved of an Announce

ptp4l.c
/**
 * @file ptp4l.c
 * @brief PTP Boundary Clock or Transparent Clock main program
 * @note Copyright (C) 2011 Richard Cochran <richardcochran@gmail.com>
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along
 * with this program; if not, write to the Free Software Foundation, Inc.,
 * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
 */
#include <limits.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>

#include "clock.h"
#include "config.h"
#include "ntpshm.h"
#include "pi.h"
#include "print.h"
#include "raw.h"
#include "sad.h"
#include "sk.h"
#include "transport.h"
#include "udp6.h"
#include "uds.h"
#include "util.h"
#include "version.h"

static void usage(char *progname)
{
	fprintf(stderr,
		"\nusage: %s [options]\n\n"
		" Delay Mechanism\n\n"
		" -A        Auto, starting with E2E\n"
		" -E        E2E, delay request-response (default)\n"
		" -P        P2P, peer delay mechanism\n\n"
		" Network Transport\n\n"
		" -2        IEEE 802.3\n"
		" -4        UDP IPV4 (default)\n"
		" -6        UDP IPV6\n\n"
		" Time Stamping\n\n"
		" -H        HARDWARE (default)\n"
		" -S        SOFTWARE\n"
		" -L        LEGACY HW\n\n"
		" Other Options\n\n"
		" -f [file] read configuration from 'file'\n"
		" -i [dev]  interface device to use, for example 'eth0'\n"
		"           (may be specified multiple times)\n"
		" -p [dev]  Clock device to use, default auto\n"
		"           (ignored for SOFTWARE/LEGACY HW time stamping)\n"
		" -s        client only synchronization mode (overrides configuration file)\n"
		" -l [num]  set the logging level to 'num'\n"
		" -m        print messages to stdout\n"
		" -q        do not print messages to the syslog\n"
		" -v        prints the software version and exits\n"
		" -h        prints this message and exits\n"
		"\n",
		progname);
}

int main(int argc, char *argv[])
{
	char *config = NULL, *req_phc = NULL, *progname;
	enum clock_type type = CLOCK_TYPE_ORDINARY;
	int c, err = -1, index, cmd_line_print_level;
	struct clock *clock = NULL;
	struct option *opts;
	struct config *cfg;

	if (handle_term_signals())
		return -1;

	cfg = config_create();
	if (!cfg) {
		return -1;
	}
	opts = config_long_options(cfg);

	/* Process the command line arguments. */
	progname = strrchr(argv[0], '/');
	progname = progname ? 1+progname : argv[0];
	while (EOF != (c = getopt_long(argc, argv, "AEP246HSLf:i:p:sl:mqvh",
				       opts, &index))) {
		switch (c) {
		case 0:
			if (config_parse_option(cfg, opts[index].name, optarg))
				goto out;
			break;
		case 'A':
			if (config_set_int(cfg, "delay_mechanism", DM_AUTO))
				goto out;
			break;
		case 'E':
			if (config_set_int(cfg, "delay_mechanism", DM_E2E))
				goto out;
			break;
		case 'P':
			if (config_set_int(cfg, "delay_mechanism", DM_P2P))
				goto out;
			break;
		case '2':
			if (config_set_int(cfg, "network_transport",
					    TRANS_IEEE_802_3))
				goto out;
			break;
		case '4':
			if (config_set_int(cfg, "network_transport",
					    TRANS_UDP_IPV4))
				goto out;
			break;
		case '6':
			if (config_set_int(cfg, "network_transport",
					    TRANS_UDP_IPV6))
				goto out;
			break;
		case 'H':
			if (config_set_int(cfg, "time_stamping", TS_HARDWARE))
				goto out;
			break;
		case 'S':
			if (config_set_int(cfg, "time_stamping", TS_SOFTWARE))
				goto out;
			break;
		case 'L':
			if (config_set_int(cfg, "time_stamping", TS_LEGACY_HW))
				goto out;
			break;
		case 'f':
			config = optarg;
			break;
		case 'i':
			if (!config_create_interface(optarg, cfg))
				goto out;
			break;
		case 'p':
			req_phc = optarg;
			break;
		case 's':
			if (config_set_int(cfg, "clientOnly", 1)) {
				goto out;
			}
			break;
		case 'l':
			if (get_arg_val_i(c, optarg, &cmd_line_print_level,
					  PRINT_LEVEL_MIN, PRINT_LEVEL_MAX))
				goto out;
			config_set_int(cfg, "logging_level", cmd_line_print_level);
			break;
		case 'm':
			config_set_int(cfg, "verbose", 1);
			break;
		case 'q':
			config_set_int(cfg, "use_syslog", 0);
			break;
		case 'v':
			version_show(stdout);
			return 0;
		case 'h':
			usage(progname);
			return 0;
		case '?':
			usage(progname);
			goto out;
		default:
			usage(progname);
			goto out;
		}
	}

	if (config && (c = config_read(config, cfg))) {
		return c;
	}

	print_set_progname(progname);
	print_set_tag(config_get_string(cfg, NULL, "message_tag"));
	print_set_verbose(config_get_int(cfg, NULL, "verbose"));
	print_set_syslog(config_get_int(cfg, NULL, "use_syslog"));
	print_set_level(config_get_int(cfg, NULL, "logging_level"));

	assume_two_step = config_get_int(cfg, NULL, "assume_two_step");
	sk_check_fupsync = config_get_int(cfg, NULL, "check_fup_sync");
	sk_tx_timeout = config_get_int(cfg, NULL, "tx_timestamp_timeout");
	sk_hwts_filter_mode = config_get_int(cfg, NULL, "hwts_filter");

	ptp_hdr_ver = config_get_int(cfg, NULL, "ptp_minor_version");
	ptp_hdr_ver = (ptp_hdr_ver << 4) | PTP_MAJOR_VERSION;

	if (sad_create(cfg)) {
		goto out;
	}

	if (config_get_int(cfg, NULL, "clock_servo") == CLOCK_SERVO_NTPSHM) {
		config_set_int(cfg, "kernel_leap", 0);
		config_set_int(cfg, "sanity_freq_limit", 0);
	}

	if (STAILQ_EMPTY(&cfg->interfaces)) {
		fprintf(stderr, "no interface specified\n");
		usage(progname);
		goto out;
	}

	type = config_get_int(cfg, NULL, "clock_type");
	switch (type) {
	case CLOCK_TYPE_ORDINARY:
		if (cfg->n_interfaces > 1) {
			type = CLOCK_TYPE_BOUNDARY;
		}
		break;
	case CLOCK_TYPE_BOUNDARY:
		if (cfg->n_interfaces < 2) {
			fprintf(stderr, "BC needs at least two interfaces\n");
			goto out;
		}
		break;
	case CLOCK_TYPE_P2P:
		if (cfg->n_interfaces < 2) {
			fprintf(stderr, "TC needs at least two interfaces\n");
			goto out;
		}
		if (DM_P2P != config_get_int(cfg, NULL, "delay_mechanism")) {
			fprintf(stderr, "P2P_TC needs P2P delay mechanism\n");
			goto out;
		}
		break;
	case CLOCK_TYPE_E2E:
		if (cfg->n_interfaces < 2) {
			fprintf(stderr, "TC needs at least two interfaces\n");
			goto out;
		}
		if (DM_E2E != config_get_int(cfg, NULL, "delay_mechanism")) {
			fprintf(stderr, "E2E_TC needs E2E delay mechanism\n");
			goto out;
		}
		break;
	case CLOCK_TYPE_MANAGEMENT:
		goto out;
	}

	clock = clock_create(type, cfg, req_phc);
	if (!clock) {
		fprintf(stderr, "failed to create a clock\n");
		goto out;
	}

	err = 0;

	while (is_running()) {
		if (clock_poll(clock))
			break;
	}
out:
	if (clock)
		clock_destroy(clock);
	sad_destroy(cfg);
	config_destroy(cfg);
	return err;
}
